YeelightAPI/DeviceLocator.cs

```diff
@@ -14,244 +16,441 @@
-        {
-            List<Task<List<Device>>> tasks = CreateDiscoverTasks(preferedInterface);
-            List<Device> devices = new List<Device>();
+    [Obsolete("Deprecated. Event will be removed in next release.")]
+    public static event DeviceFoundEventHandler OnDeviceFound;
 
-            if (tasks.Count != 0)
-            {
-                await Task.WhenAll(tasks);
+    /// <summary>
+    ///   Notification Received event handler
+    /// </summary>
+    /// <param name="sender"></param>
+    /// <param name="e"></param>
+    [Obsolete("Deprecated. Event will be removed in next release.")]
+    public delegate void DeviceFoundEventHandler(object sender, DeviceFoundEventArgs e);
 
-                devices.AddRange(tasks.SelectMany(t => t.Result).GroupBy(d => d.Hostname).Select(g => g.First()));
-            }
+    #region Public Methods
 
-            return devices;
-        }
+    /// <summary>
+    ///   Discover devices in a specific Network Interface
+    /// </summary>
+    /// <param name="preferedInterface"></param>
+    /// <returns></returns>
+    [Obsolete("Deprecated. Use DiscoverAsync and overloads instead.")]
+    public static async Task<List<Device>> Discover(NetworkInterface preferedInterface)
+    {
+      List<Task<List<Device>>> tasks = DeviceLocator.CreateDiscoverTasks(preferedInterface);
 
-        /// <summary>
-        /// Discover devices in LAN
-        /// </summary>
-        /// <returns></returns>
-        public static async Task<List<Device>> Discover()
-        {
-            List<Task<List<Device>>> tasks = new List<Task<List<Device>>>();
-            List<Device> devices = new List<Device>();
+      List<Device>[] result = await Task.WhenAll(tasks);
+      return result.SelectMany(devices => devices).ToList();
+    }
 
-            foreach (NetworkInterface ni in NetworkInterface.GetAllNetworkInterfaces().Where(n => n.OperationalStatus == OperationalStatus.Up))
-            {
-                tasks.AddRange(CreateDiscoverTasks(ni));
-            }
+    /// <summary>
+    ///   Discover devices in LAN
+    /// </summary>
+    /// <returns></returns>
+    [Obsolete("Deprecated. Use DiscoverAsync and overloads instead.")]
+    public static async Task<List<Device>> Discover()
+    {
+      var tasks = new List<Task<List<Device>>>();
 
-            if (tasks.Count != 0)
-            {
-                await Task.WhenAll(tasks);
+      foreach (NetworkInterface ni in NetworkInterface.GetAllNetworkInterfaces()
+        .Where(n => n.OperationalStatus == OperationalStatus.Up))
+      {
+        tasks.AddRange(DeviceLocator.CreateDiscoverTasks(ni));
+      }
 
-                devices.AddRange(tasks.SelectMany(t => t.Result).GroupBy(d => d.Hostname).Select(g => g.First()));
-            }
 
-            return devices;
-        }
+      List<Device>[] result = await Task.WhenAll(tasks);
+      return result.SelectMany(devices => devices).ToList();
```

Comment:
>Even if this method is now obsolete, we need to keep the previous code to ensure unicity of found devices : 
`return result.SelectMany(devices => devices).`**GroupBy(d => d.Hostname).Select(g => g.First())**`.ToList()`

Same thing here

Reply:
Fixed. I've also added the grouping to the new `DiscoverAsync`. 

I also refactored the `CheckSocketForDevices` method (extracted to new method).